torch/_higher_order_ops/scan.py

```diff
@@ -326,9 +354,236 @@ def scan_op_dense(combine_fn, input, init, dim):
     return generic_scan(combine_fn, input, init, dim)
 
 
-scan_op.py_impl(DispatchKey.Autograd)(
-    autograd_not_implemented(scan_op, deferred_error=True)
-)
+class ScanAutogradOp(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        fw_graph,
+        joint_graph,
+        dim,
+        num_leaves,
+        *ops,
+    ):
+        init = ops[:num_leaves]
+        input = ops[num_leaves:]
+
+        ctx._joint_graph = joint_graph
+        ctx._dim = dim
+        ctx._num_leaves = num_leaves
+        ctx._num_elems = input[0].size()[dim] + 1
+
+        outs = generic_scan(fw_graph, input, init, dim)
+        input_init = [
+            torch.concatenate([ini, inp], dim=dim) for ini, inp in zip(init, input)
+        ]
+        ctx.save_for_backward(*(input_init + list(outs)))
+
+        return tuple(outs)
+
+    @staticmethod
+    def backward(ctx, *flat_grads):
```

Comment:
The backward of scan should be another scan operator. Here is how to construct the operator:
Consider the forward:
```python
# scan(f, init=init, inputs=[x0, x1, x2])
carry_0, y0 = f(init, x0)
carry_1, y1 = f(carry_0, x1)
carry_2, y2 = f(carry_1, x2)
# The return of scan operator is:
carry_2, [y0, y1, y2]
```
Now consider the backward of the scan, we're given the gradient of scan's forward output i.e. g_carry_2, [gy0, gy1, gy2] and we can compute the griadient of input i.e. g_init and [gx0, gx1, gx2] by:
```python
g_carry_1, gx2 = gf(g_carry_2, gy2, carry_1, x2)
g_carry_0, gx1 = gf(g_carry_1, gy1, carry_0, x1)
g_init, gx0 = gf(g_carry_0, gy0, init, x0)
```
We find that the above can be expressed as a scan operator as below:
```
g_init, g_x0 = scan(
  gf, 
  init=g_carry_2,
  inputs=([gy0, gy1, gy2], [init, carry_0, carry_1], [x0, x1, x2]),
  reverse=True
)
```